### ISSUE TITLE
[HOLD for payment 2024-01-24] CRITICAL: [$750] Default new rooms to your primary workspace

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#32435]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.7-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1701632317613889

## Action Performed:
1. Click fab and select start chat

## Expected Result:
workspace field should default to primary workspace

## Actual Result:
workspace field is blank

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![image (5)](https://github.com/Expensify/App/assets/38435837/7d142909-c10f-4215-9d49-8e6eeacf8528)

![Snip - (57) New Expensify - Google Chrome](https://github.com/Expensify/App/assets/38435837/e63688c8-850b-45eb-b88c-7b6c5b69f8bf)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b604efe47660d3a4</li>
        <li>Upwork Job ID: 1731962570569764864</li>
        <li>Last Price Increase: 2024-01-08</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 28003794</li>
<li>esh-g | Contributor | 28003795</li>
</ul></li>
    </ul>
</details>

**Comment by esh-g** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.
When creating a Room, workspace should default to primary workspace

## What is the root cause of that problem?
Feature request. N/A

## What changes do you think we should make in order to solve the problem?
We should do the following changes:
1. Make the workspace value picker a controlled input by adding the `value={policyID}` attribute to it here:
https://github.com/Expensify/App/blob/6dbaa4a7ac8c8142a452bf2a059a16b2947616c9/src/pages/workspace/WorkspaceNewRoomPage.js#L272-L277
2. Modify the ValuePicker to prevent a `null` value from causing an error here by using optional chaining(`value?.length`):
https://github.com/Expensify/App/blob/6dbaa4a7ac8c8142a452bf2a059a16b2947616c9/src/components/ValuePicker/index.js#L65
3. Add a `useEffect` to `WorkspaceNewRoomPage` to set the `policyID` to the first workspace:
```js
useEffect(() => {
    setPolicyID(workspaceOptions[0].value);
}, [props.policies]);
```

## What alternative solutions did you explore? (Optional)
If we don't want to use the first element, we can possibly make the last modified workspace as the default selected which can be done by refactoring this method:
https://github.com/Expensify/App/blob/6dbaa4a7ac8c8142a452bf2a059a16b2947616c9/src/libs/PolicyUtils.ts#L16-L20
to something like this:
```js
function getActivePolicies(policies: OnyxCollection<Policy>): Policy[] | undefined {
    const activePolicies = Object.values(policies ?? {}).filter<Policy>(
        (policy): policy is Policy =>
            policy !== null && policy && (policy.isPolicyExpenseChatEnabled || policy.areChatRoomsEnabled) && policy.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE,
    );
    activePolicies.sort((a, b) => Number(b.lastModified) - Number(a.lastModified));
    return activePolicies;
}
```

**Comment by dylanexpensify** [comment]
Woot, welcome to the jungle @alitoshmatov! We already have a proposal from @dukenv0307 to review! 🎉 Can we make sure we review it today? 🙇‍♂️

**Comment by alitoshmatov** [comment]
Yes. I will review it today

**Comment by yh-0218** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When creating a Room, workspace should default to primary workspace
### What is the root cause of that problem?
https://github.com/Expensify/App/blob/48b5a29f33ca4cfa8a0b75361f68f6df83455613/src/pages/workspace/WorkspaceNewRoomPage.js#L79
Because we set `policyID = null` and reset from only workspace list selecting.
### What changes do you think we should make in order to solve the problem?
1. we need to init `policyID`
```
     useEffect(() => {
            if (workspaceOptions && workspaceOptions.length > 0) {
                setPolicyID(workspaceOptions[0].value);  // We can update logic
            } else {
                setPolicyID(null);
            }
      }, [workspaceOptions])
```
2. we need to set `policyID` to component as value
`value={policyID}` to here
https://github.com/Expensify/App/blob/48b5a29f33ca4cfa8a0b75361f68f6df83455613/src/pages/workspace/WorkspaceNewRoomPage.js#L217
3. update one more
```
const descStyle = value && value.length === 0 ? StyleUtils.getFontSizeStyle(variables.fontSizeLabel) : null;
```
https://github.com/Expensify/App/blob/48b5a29f33ca4cfa8a0b75361f68f6df83455613/src/components/ValuePicker/index.js#L65
### What alternative solutions did you explore? (Optional)


https://github.com/Expensify/App/assets/65789015/083a5a8d-9f55-4ad7-be8c-42def81e262e

**Comment by alitoshmatov** [comment]
@dukenv0307 Your solution won't update initial value of ValuePicker.

**Comment by alitoshmatov** [comment]
@dukenv0307 I much prefer to turn ValuePicker into controlled component, I don't think providing defaultValue prop separately and setting default value for `policyID` separately is a good idea. We are much better off providing value prop to ValuePicker component

**Comment by alitoshmatov** [comment]
@esh-g 's [proposal](https://github.com/Expensify/App/issues/32435#issuecomment-1839391391) looks great and solve the issue.

We can go with that.

C+ reviewed 🎀 👀 🎀

**Comment by esh-g** [comment]
@alitoshmatov Would you like to wait for the active workspace to get implemented before we work on this? (Described in the alternate solution section of the proposal)

**Comment by alitoshmatov** [comment]
@dukenv0307 I understand what you mean, but your proposal wasn't clear enough. I would have understood if you had missed small detail, but it wasn't. In the first look, I didn't even assumed you meant this.

**Comment by alitoshmatov** [comment]
Can you guys help what primary workspace mean, do we have logic to determine that?

Or should we wait for https://github.com/Expensify/App/issues/31887 as @esh-g suggested?

cc: @dylanexpensify @AndrewGable

**Comment by alitoshmatov** [comment]
> @alitoshmatov I consider about `workspaceOptions.length`

I am doubtful about this approach, can we make sure this is the best approach and doesn't cause any regressions

**Comment by alitoshmatov** [comment]
> Modify the ValuePicker to prevent a null value from causing an error here by using optional chaining(value?.length):
@esh-g 
I am doubtful about this approach, can we make sure this is the best approach and doesn't cause any regressions

**Comment by esh-g** [comment]
@alitoshmatov I have tested this and it seems to work as intended...

<img width="216" alt="Screenshot 2023-12-06 at 9 45 38 PM" src="https://github.com/Expensify/App/assets/77237602/16ec4b82-9911-4a17-ba16-2086a575896e">

but if you're still doubtful, maybe we can just do `value && value.length === 0` only a minor change but same gist.

The main question is that should we wait for the `ACTIVE_WORKSPACE_ID` implementation, which is what I think we should do.

**Comment by alitoshmatov** [comment]
@dylanexpensify We can move forward with assigning @esh-g . One thing holding us this concern:

> Can you guys help what primary workspace mean, do we have logic to determine that?
Or should we wait for https://github.com/Expensify/App/issues/31887 as @esh-g suggested?

**Comment by dylanexpensify** [comment]
Assigned esh-g! @AndrewGable can you confirm outstanding query? Unsure the exact logic we use to determine it (likely just the most recently used one, or the only one available I'm guessing)

**Comment by quinthar** [comment]
Every account has a "primary policy" -- this is the policy/workspace that gets receipts from receipts@expensify.com.  I think we should use that one.

**Comment by dylanexpensify** [comment]
Fantastic, thanks for confirming @quinthar! @alitoshmatov @esh-g does that sound good?

**Comment by esh-g** [comment]
Yes, I think that means we will have to wait for the active workspace implementation before we work on this, right? @alitoshmatov

**Comment by dylanexpensify** [comment]
@alitoshmatov @AndrewGable can you confirm?

**Comment by alitoshmatov** [comment]
I couldn't find any information about primary policy. @esh-g I don't think @quinthar is implying this new feature https://github.com/Expensify/App/issues/31887. 

@AndrewGable Can you jump in to clarify this

**Comment by dylanexpensify** [comment]
Bump @AndrewGable 🙇‍♂️

**Comment by AndrewGable** [comment]
Yes, I don't know if a default policy exists on NewDot yet. Can we unwind any proposals until we have this confirmed? Thanks!

**Comment by esh-g** [comment]
Yes @AndrewGable I've mentioned that it will get added here https://github.com/Expensify/App/issues/31887

**Comment by AndrewGable** [comment]
At minimum yes, the internal discussion is ongoing.

**Comment by dylanexpensify** [comment]
@AndrewGable mind posting where we landed?

**Comment by roryabraham** [comment]
I suppose this is in my court now. The planned back-end changes, for those who missed the internal slack convo, is to bring an existing piece of back-end data, `activePolicyID`, to the NewDot front-end. This will be a field under the `ACCOUNT` Onyx key.

**Comment by roryabraham** [comment]
More detail, mostly as a note-to-self:

- Update `OpenApp` and `ReconnectApp` Auth commands to include Onyx instructions to set the `activePolicyID` in Auth
- Hook into `NVPStore::postProcess`, and if `activePolicyID` was set, then use PHP to send Onyx instructions to update the `activePolicyID`.

**Comment by roryabraham** [comment]
Auth PR is ready: https://github.com/Expensify/Auth/pull/9559

**Comment by roryabraham** [comment]
Web-E PR is ready: https://github.com/Expensify/Web-Expensify/pull/40310

**Comment by roryabraham** [comment]
PR to update Account Onyx type in this repo: https://github.com/Expensify/App/pull/33728

**Comment by roryabraham** [comment]
So currently we have three PRs ready for review:

- https://github.com/Expensify/Auth/pull/9559
- https://github.com/Expensify/Web-Expensify/pull/40310
- https://github.com/Expensify/App/pull/33728

They can be safely merged + deployed in any order, and together they should reliably add the `activePolicyID` property to the account object in Onyx.

_Caveat:_ The only exception is that `activePolicyID` might not always be up-to-date for the relatively small subset of users that are not yet on the reliable updates beta. This caveat was discussed [here](https://expensify.slack.com/archives/C03SDMF9YJ2/p1703802422507079?thread_ts=1703801877.518519&cid=C03SDMF9YJ2)

**Comment by roryabraham** [comment]
Oh, and also I want to clarify something that was discussed internally to clarify what we mean by "primary workspace".

Really there are two separate pieces of data:

- `activePolicyID`: This is an existing piece of data that's used in the back-end to determine which policy to attach a new expense to when generating an expense via SmartScan (e.g: forwarding an email receipt to receipts@expensify.com). This is a persistent piece of data that, after the above PRs are merged, will be included in the Account object in Onyx.
- `selectedWorkspace` (this is a working name, actual name TBD): This is a piece of data that will be introduced in https://github.com/Expensify/App/issues/31887. It is front-end-only, and represents which workspace the user has selected / is "filtered on". It _may_ end up living in Onyx, or it may end up being encoded in the URL somehow. I'm not entirely sure. However, it won't be stored in the in the database or be persisted if a user logs out.

So when generating a new workspace, or creating an expense, or potentially in other flows as well, the logic for picking a policy will generally be as follows:

- When in the Expensify “all” view, use the `activePolicyID` workspace as the default
- When “filtered” to a specific workspace view, use the `selectedWorkspace` as the most contextually relevant workspace to populate when going to create a new room / new expense / whatever.
- The “workspace” field in the new room creation flow can be changed in both cases

**Comment by roryabraham** [comment]
So as soon as all three of the above PRs are deployed to prod, we can start using the `activePolicyID`. When `selectedWorkspace` and the workspace switcher is built, we can start using that first, then falling back on `activePolicyId`

**Comment by roryabraham** [comment]
So @esh-g, since you've been hired for this job, I'd like to task you with completing an implementation similar to your original proposal – just instead of using the first workspace, use the workspace with the ID matching the `activePolicyID` that will soon be added to the Account in Onyx.

Sound good? If you agree, please react with a 👍🏼 and then I'll notify you when all the back-end pieces have been deployed and you can proceed. If you're no longer interested, please react with a 👎🏼 and I'll unassign you. Thanks, and if you have any questions don't hesitate to ask!

**Comment by esh-g** [comment]
@roryabraham sounds great 👍 exactly what I had in my mind as well!

**Comment by quinthar** [comment]
Just to confirm, at this point are we waiting for @roryabraham to deploy the server fixes before @esh-g can start?  Or what's the status?

**Comment by esh-g** [comment]
Yes @quinthar exactly! We need to wait for the primary workspace implementation

**Comment by roryabraham** [comment]
Thanks for the patience everyone. A few of the activePolicyID PRs have been deployed, I hit a minor snag with some tests which may delay us by a day or two. Next steps are:

- https://github.com/Expensify/Web-Expensify/pull/40325 must be merged and deployed to prod
- Then https://github.com/Expensify/Auth/pull/9559 must be merged and deployed to prod

Then @esh-g will be unblocked here. I will have to follow-up with another PR in Web-Expensify to re-enable some temporarily disabled tests.

---

@esh-g, if you want to get a head-start, I notice that unit tests covering `Policy::createWorkspace` are woefully absent. If you invest some time in writing unit tests to cover those: namely, covering the API call, optimisticData, successData, and failureData, we can pay a $250 bonus for that. Let me know, thanks!

**Comment by esh-g** [comment]
@roryabraham Sure! Could you once please refer to a test written for an another API command so that I can have some reference for the Expensify style as well?

**Comment by roryabraham** [comment]
The last back-end PR was merged earlier today. We usually don't deploy on Fridays so everything will be on prod by EOD Monday.

**Comment by esh-g** [comment]
@roryabraham I created a draft PR with the test for `Policy::createWorkspace` here: https://github.com/Expensify/App/pull/34023

Could you please raise any concerns that might be present?

**Comment by roryabraham** [comment]
Left some feedback. Overall looks great so far!

**Comment by dylanexpensify** [comment]
Nice! @esh-g mind giving us a new ETA?

**Comment by esh-g** [comment]
@roryabraham I think I found a problem in the current codebase while writing the test...

In the `successData` of `Policy::createWorkspace` while setting the `pendingAction: null` for the `CREATED` report action for admin, expense and announce report, we use the `reportActionID` incorrectly here:
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1308-L1310
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1327-L1329
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1346-L1348

<img width="410" alt="Screenshot 2024-01-08 at 7 04 45 PM" src="https://github.com/Expensify/App/assets/77237602/cb6bf73a-269a-43ff-a2ca-5fbf95779bc0">

We are using the first key as the `reportActionID` but in actual, the `adminChatData` is the just the `report` object itself not a `rNVP`. This means we are setting `type: {pendingAction: null}` which is obviously wrong.
This is causing my tests to fail...

We need to use the `chatReportActionIDs` for each report here:
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1177
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1181
https://github.com/Expensify/App/blob/288beccda639c96895c5af56695877a8a42da731/src/libs/actions/Policy.js#L1185

@dylanexpensify Should I raise another PR to fix this in the codebase or include it in the test PR itself?

**Comment by esh-g** [comment]
> Nice! @esh-g mind giving us a new ETA?

As for ETA, I have the tests refactored and also fixed the above bug in my PR (also marked it as ready for review)
So, I am also waiting on getting the `activeWorkspaceID` code implemented so that I can write a test for that and use it to fix the actual bug as well.

**Comment by roryabraham** [comment]
Great find @esh-g! I was thinking that since the `activeWorkspaceID` code won't be deployed until later today, we could merge your PR that introduces the base test and do a second PR to actually implement the changes outlined in this issue, along with the updates in the test that show the changed behavior. I think that could end up being cleaner.

@esh-g what do you think?

**Comment by esh-g** [comment]
Sure @roryabraham I'll be available to raise the second PR as well!

**Comment by roryabraham** [comment]
Great, I left a few minor comments on the test PR, then I think we can merge that one and do a second follow-up PR to finish this issue out.

**Comment by roryabraham** [comment]
Upping the bounty for this issue per [this comment](https://github.com/Expensify/App/issues/32435#issuecomment-1873446985)

**Comment by roryabraham** [comment]
Merged the PR to introduce tests 🚀

**Comment by roryabraham** [comment]
@esh-g all the back-end pieces are deployed, so E/App should have an `activePolicyID` field under the `Account` object in Onyx now. Looking forward to seeing the main PR discussed [above](https://github.com/Expensify/App/issues/32435#issuecomment-1871678744) ASAP. Thanks!

**Comment by esh-g** [comment]
gentle bump @roryabraham

**Comment by roryabraham** [comment]
> Is the activePolicyID always set in the backend?

Currently it's controlled in the policy settings editor in OldDot and always set when creating a policy on expensify.com, but I don't know if it's set for new workspaces in NewDot. Let me check. If not another 1 or 2 back-end PRs may be needed.

However, that doesn't need to block this issue to default new rooms to your primary workspace. If you need to make sure `activePolicyID` is set, sign into https://staging.expensify.com (instead of https://staging.new.expensify.com) with your test account, go to Settings -> Workspaces -> settings cog on the workspace you want to make your primary -> make primary

> Also what tests should I add in the policyTests for the activePolicyID?

I suppose the main thing we would want to test in this is actually a UI thing – that when you open the new room page from the global create menu that the workspace is pre-populated by your `activePolicyID`. I don't think that fits in the test you already created though.

I'll get back to you if there's anything we want to add to the policy test you created.

**Comment by esh-g** [comment]
@roryabraham As per my understanding, then I don't need to add any additional test for now right? And could you please point out any other thing that might be missing in the PR?

**Comment by roryabraham** [comment]
Yeah, I think no additional tests necessary for now

**Comment by dylanexpensify** [comment]
coming up!

**Comment by dylanexpensify** [comment]
tomorrow!

**Comment by alitoshmatov** [comment]
Regression Test Proposal

1. Make sure you have a workspace
2. If you don't have a workspace create one and logout and login again
3. Click on FAB
4. Click on `Start chat`
5. Choose `Room` tab
6. Make sure you have a default workspace selected on workspace selection

Do we agree 👍 or 👎

**Comment by dylanexpensify** [comment]
Payment summary:
- Contributor: @esh-g $750
- Contributor+: @alitoshmatov $750

Please apply or request!

**Comment by dylanexpensify** [comment]
done
